File: src/lira/linguistics/data/sentence.py
"""The top-level read/write unit: one or more Clauses plus terminal
punctuation (Linguistics Layer developer specification, 6; spec 14.1).
References Vocabulary Words -- never copies or replaces their lexical
data (Rule 17). Word and PartOfSpeech are used only as type hints here,
left unimported at module scope -- same import-cycle reasoning as
clause.py."""

from dataclasses import dataclass, field
from typing import List, Optional, Sequence, Tuple, Union

from .clause import Clause
from .interpretation import Interpretation
from .linguistic_unit import LinguisticUnit
from .reading_error import ReadingError
from .sentence_type import SentenceType
from .validation_outcome import ValidationOutcome


@dataclass
class Sentence(LinguisticUnit):
    clauses: List[Clause] = field(default_factory=list, kw_only=True)
    requires_punctuation: Optional[bool] = field(default=None, kw_only=True)

    tokens: List["Word"] = field(default_factory=list, kw_only=True)
    sentence_type: Optional[SentenceType] = field(default=None, kw_only=True)
    selected_parts_of_speech: Tuple["PartOfSpeech", ...] = field(default=(), kw_only=True)
    punctuation: Optional["Word"] = field(default=None, kw_only=True)
    validation: ValidationOutcome = field(default=ValidationOutcome.UNRESOLVED, kw_only=True)
    confidence: float = field(default=0.0, kw_only=True)
    alternatives: Tuple[Interpretation, ...] = field(default=(), kw_only=True)
    errors: Tuple[ReadingError, ...] = field(default=(), kw_only=True)
    # system_property inherited from LinguisticUnit -- allocated by
    # GraphProcessor at materialisation time, not here (see phrase.py's
    # own note on the same point).

    @classmethod
    def read(
        cls,
        text_or_tokens: Union[str, Sequence["TokenReading"]],
        *,
        context: "ReadingContext",
        grammar: Optional["GrammarConfigurator"] = None,
    ) -> "Sentence":
        """Spec 14.3 entry point ("accepts text or pre-resolved
        tokens"). Raw text is tokenised and resolved via
        context.token_resolver; an already-resolved TokenReading
        sequence is read as-is -- the latter is what
        LinguisticController.read_text uses internally when handing
        SentenceReader one sentence's slice of a larger, already-
        tokenised prompt. Contains no grammar or sequencing logic of its
        own (spec 9) -- one delegation to the shared SentenceReader,
        reached through `context` rather than a controller-shaped
        dependency (see role/reading_context.py's own docstring for
        why)."""
        return context.sentence_reader.read(
            text_or_tokens,
            grammar=grammar or context.grammar,
        )

File: prototype/src/lira/linguistics/data/sentence.ts
import type { PartOfSpeech } from "../../vocabulary/data/enums/part_of_speech";
import type { Word } from "../../vocabulary/data/word";
import type { GrammarConfigurator } from "../role/grammar_configurator";
import type { ReadingContext } from "../role/reading_context";
import type { Clause } from "./clause";
import type { Interpretation } from "./interpretation";
import type { LinguisticUnit } from "./linguistic_unit";
import type { ReadingError } from "./reading_error";
import type { SentenceType } from "./sentence_type";
import type { TokenReading } from "./token_reading";
import { ValidationOutcome } from "./validation_outcome";

/** The top-level read/write unit: one or more Clauses plus terminal
 * punctuation (Linguistics Layer developer specification, 6; spec
 * 14.1). References Vocabulary Words -- never copies or replaces their
 * lexical data (Rule 17).
 *
 * Ported from linguistics/data/sentence.py. */
export interface Sentence extends LinguisticUnit {
  clauses: Clause[];
  requiresPunctuation?: boolean;

  tokens: Word[];
  sentenceType?: SentenceType;
  selectedPartsOfSpeech: readonly PartOfSpeech[];
  punctuation?: Word;
  validation: ValidationOutcome;
  confidence: number;
  alternatives: readonly Interpretation[];
  errors: readonly ReadingError[];
}

export function createSentence(init: Pick<Sentence, "text"> & Partial<Sentence>): Sentence {
  return {
    clauses: [],
    tokens: [],
    selectedPartsOfSpeech: [],
    validation: ValidationOutcome.UNRESOLVED,
    confidence: 0.0,
    alternatives: [],
    errors: [],
    ...init,
  };
}

export interface ReadSentenceOptions {
  grammar?: GrammarConfigurator;
}

/** Spec 14.3 entry point ("accepts text or pre-resolved tokens"). Raw
 * text is tokenised and resolved via context.tokenResolver; an already-
 * resolved TokenReading sequence is read as-is -- the latter is what
 * LinguisticController.readText uses internally when handing
 * SentenceReader one sentence's slice of a larger, already-tokenised
 * prompt. Contains no grammar or sequencing logic of its own (spec 9)
 * -- one delegation to the shared SentenceReader. */
export function readSentence(
  textOrTokens: string | readonly TokenReading[],
  context: ReadingContext,
  options: ReadSentenceOptions = {},
): Sentence {
  return context.sentenceReader.read(textOrTokens, { grammar: options.grammar ?? context.grammar });
}
